azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,698 @@
+
+    @staticmethod
+    def track_cgroup(cgroup):
+        """
+        Create a CGroupsTelemetry object to track a particular CGroups instance. Typical usage:
+        1) Create a CGroups object
+        2) Ask CGroupsTelemetry to track it
+        3) Tell the CGroups object to add one or more processes (or let systemd handle that, for its cgroups)
+
+        :param CGroups cgroup: The cgroup to track
+        """
+        name = cgroup.name
+        if not CGroupsTelemetry.is_tracked(name):
+            tracker = CGroupsTelemetry(name, cgroup=cgroup)
+            CGroupsTelemetry._tracked[name] = tracker
+
+    @staticmethod
+    def track_systemd_service(name):
+        """
+        Create the CGroups object for a systemd service and track it.
+
+        :param str name: Service name (without .service suffix) to be tracked.
+        """
+        service_name = "{0}.service".format(name)
+        if not CGroupsTelemetry.is_tracked(service_name):
+            cgroup = CGroups.for_systemd_service(service_name)
+            tracker = CGroupsTelemetry(service_name, cgroup=cgroup)
+            CGroupsTelemetry._tracked[service_name] = tracker
+
+    @staticmethod
+    def track_extension(name, cgroup=None):
+        """
+        Create all required CGroups to track all metrics for an extension and its associated services.
+
+        :param str name: Full name of the extension to be tracked
+        :param CGroups cgroup: CGroup for the extension itself. This method will create it if none is supplied.
+        """
+        if not CGroupsTelemetry.is_tracked(name):
+            cgroup = CGroups.for_extension(name) if cgroup is None else cgroup
+            logger.info("Now tracking cgroup {0}".format(name))
+            CGroupsTelemetry.track_cgroup(cgroup)
+        if CGroups.is_systemd_manager():
+            if name in related_services:
+                for service_name in related_services[name]:
+                    CGroupsTelemetry.track_systemd_service(service_name)
+
+    @staticmethod
+    def is_tracked(name):
+        return name in CGroupsTelemetry._tracked
+
+    @staticmethod
+    def stop_tracking(name):
+        """
+        Stop tracking telemetry for the CGroups associated with an extension. If any system services are being
+        tracked, those will continue to be tracked; multiple extensions might rely upon the same service.
+
+        :param str name: Extension to be dropped from tracking
+        """
+        if CGroupsTelemetry.is_tracked(name):
+            del (CGroupsTelemetry._tracked[name])
+
+    @staticmethod
+    def set_delete_empty_groups(setting):
+        CGroupsTelemetry.delete_empty_groups = setting
+
+    @staticmethod
+    def collect_all_tracked():
+        """
+        Return a dictionary mapping from the name of a tracked cgroup to the list of collected metrics for that cgroup.
+
+        :returns: Dictionary of collected metrics, by cgroup
+        :rtype: dict(str: [(str, str, float)])
```

Comment:
You have a great comment below about what this tuple means.  I was able to figure out from previous code, but it would be nice to see "(metric group name, metric name, metric value)".

Reply:
OK